Feature/kas 2430 gebruik vergaderactiviteit codelist uit backend om keuzelijst op te stellen in frontend

sergiofenoll opened the pull request:
Adds a code list for vergaderactiviteiten.

The data stored [here](https://github.com/kanselarij-vlaanderen/frontend-kaleidos/blob/development/app/utils/config.js#L64) is now pulled in from the backend. Wherever kind URIs were used, now Ember Data records are used (meeting now has a has-one relationship to meeting-kind, Utils::KindSelector uses records instead of URIs now)

A custom adapter for meeting-kind models was added to prevent multiple requests to the backend being sent.

sergiofenoll committed:
Update transition code to remove deprecation warning

sergiofenoll committed:
Add meeting kind model to frontend

sergiofenoll committed:
Add isAnnexMeeting getter to meeting-kind model

sergiofenoll committed:
Directly use meeting-kind in meeting model

sergiofenoll committed:
Update kind-selector to use meeting-kind record

sergiofenoll committed:
Ensure meeting-kinds are only loaded in once

sergiofenoll commented:
Custom jenkins build: http://kal-kastaar.s.redpencil.io:8080/job/kaleidos/job/custom_frontend_backend_combination/171/

sergiofenoll committed:
Use .get() in case the object being accessed is a proxy

sergiofenoll committed:
Merge branch 'development' of https://github.com/kanselarij-vlaanderen/frontend-kaleidos into feature/KAS-2430-gebruik-vergaderactiviteit-codelist-uit-backend-om-keuzelijst-op-te-stellen-in-frontend

sergiofenoll committed:
Use initializeEmptyKind argument & checek for falsiness of kind

sergiofenoll committed:
Replace unnecessary async with sync function

sergiofenoll committed:
Replace missed kindToShow with kind

sergiofenoll committed:
Merge branch 'development' into feature/KAS-2430-gebruik-vergaderactiviteit-codelist-uit-backend-om-keuzelijst-op-te-stellen-in-frontend

sergiofenoll committed:
Use peekAll in KindSelector, don't try to load in meeting kinds

sergiofenoll committed:
Move preloading meeting kinds to application route

sergiofenoll commented:
Custom jenkins build: http://kal-kastaar.s.redpencil.io:8080/job/kaleidos/job/custom_frontend_backend_combination/177/

ValenberghsSven commented on app/components/agenda/printable-agenda.hbs:
```diff
@@ -18,7 +18,7 @@
       }}
       -
-      {{@meeting.kindToShow.altLabel}}
+      {{if @meeting.kind.altLabel @meeting.kind.altLabel @meeting.kind.label}}
```
Dit doen we op 3 plaatsen.
Beter een getter in het model voor minder duplicatie + comment dat `altLabel` kan leeg zijn?
Het onstaan van altLabel is er gekomen omdat de gebruikers een andere tekst vroegen voor de titels van vb agenda afdruk en titels nieuwsbrief enzo.
Uiteindelijk bleek dat er enkel een verschil was bij de elektronische procedure

ValenberghsSven commented on app/components/utils/kind-selector.hbs:
```diff
@@ -1,22 +1,13 @@
-{{! template-lint-disable no-action }}
 <div data-test-utils-kind-selector class="auk-form-group">
-  {{#if (not this.hideLabel)}}
+  {{#if (not @hideLabel)}}
```
is `hideLabel` nog nodig? Lijkt niet meer gebruikt.

ValenberghsSven commented on app/components/utils/kind-selector.js:
```diff
@@ -1,29 +1,31 @@
-  }),
+    this.kind = this.args.kind;
+    if (this.args.initializeEmptyKind && !this.args.kind) {
+      this.setAction(this.options.firstObject);
```
Ik snap de firstObject approach.
Wat je hier wel beetje verliest is dat het specifiek het type "Ministerraad" is dat de default moet zijn.
De reden dat het hier werkt, is omdat we hogerop de data hebben ingeladen met sortering op priorities (denk ik?)
Maar die sortering gaan we hier niet afdwingen, dus die `firstObject` zou ooit kunnen veranderen?
eventueel `this.store.peekAll('meeting-kind').sortBy('priority');` of position, als je backend PR bekijkt.

Ook een vermelding dat de "Ministerraad" de default is "to initialize the dropdown to the **default** meeting kind of no meeting kind was provided" kan later verwarring voorkomen.

ValenberghsSven replied:
"kind **of** no meeting kind was provided"  => "if" ?

ValenberghsSven commented on app/config/constants.js:
```diff
@@ -90,5 +90,12 @@ export default {
+  MEETING_KINDS: {
+    MINISTERRAAD: 'http://kanselarij.vo.data.gift/id/concept/ministerraad-type-codes/A5D6B7A8-2F9C-44B6-B3BE-98D80B426254',
+    ELEKTRONISCHE_PROCEDURE: 'http://kanselarij.vo.data.gift/id/concept/ministerraad-type-codes/406F2ECA-524D-47DC-B889-651893135456',
+    BIJZONEDERE_MINISTERRAAD: 'http://kanselarij.vo.data.gift/id/concept/ministerraad-type-codes/7D8E35BE-E5D1-494F-B5F9-51B07875B96F',
```
typefout `BIJZONDERE_MINISTERRAAD`

ValenberghsSven commented on app/config/constants.js:
```diff
@@ -90,5 +90,12 @@ export default {
+    BIJZONEDERE_MINISTERRAAD: 'http://kanselarij.vo.data.gift/id/concept/ministerraad-type-codes/7D8E35BE-E5D1-494F-B5F9-51B07875B96F',
+    PVV: 'http://kanselarij.vo.data.gift/id/concept/ministerraad-type-codes/1d16cb70-0ae9-489e-bf97-c74897222e3c',
+    ANNEX: 'http://kanselarij.vo.data.gift/id/concept/ministerraad-type-codes/d36138a9-07f0-4df6-bbf0-abd51a24e4ce',
+  },
```
`MINISTERRAAD_TYPES` bestaat nog in `utils/config.js`. mag weg?
https://github.com/kanselarij-vlaanderen/frontend-kaleidos/blob/feature/KAS-2430-gebruik-vergaderactiviteit-codelist-uit-backend-om-keuzelijst-op-te-stellen-in-frontend/app/utils/config.js#L64

ValenberghsSven commented on app/models/meeting.js:
```diff
@@ -108,19 +105,8 @@ export default Model.extend({
-  }),
-
   isAnnex: computed('kind', function() {
-    return isAnnexMeetingKind(this.kind);
+    return this.get('kind').get('isAnnexMeeting');
   }),
```
Deze computed wordt enkel gebruik in `new-meeting-modal.js`, deze computed weg en lokaal die code doen?

ValenberghsSven replied:
Je kan daar gewoon de lokale `!this.isAnnexMeeting` gebruiken lijkt me, we willen gewoon weten wanneer we dat verslag van de vorige vergadering moeten maken.

ValenberghsSven commented on app/components/meeting/new-meeting-modal.js:
```diff
@@ -36,7 +34,7 @@ export default class MeetingNewMeetingModal extends Component {
 
   get isAnnexMeeting() {
-    return isAnnexMeetingKind(this.selectedKindUri);
+    return this.kind?.get('isAnnexMeeting');
```
moet deze `get()` of kunnen we gewoon `this.kind?.isAnnexMeeting`?

ValenberghsSven replied:
Dit komt omdat we `this.broader` gaan bekijken vermoed ik? Daardoor zit je met een proxy en heb je die get nodig.
Kunnen we dat anders oplossen? 
- broader al includen in de query in application route?
- broader ophalen ergens anders?
- uri van pvv kind checken ipv de broader relatie? (is dan synchroon, zou enkel problemen geven als er later extra types annex bekomen, wat ik betwijfel)

sergiofenoll replied:
Die `.get()` is inderdaad niet meer nodig. Aanvankelijk was `broarder` een relatie ipv een attribuut, waardor het wel nodig was. Maar om het simpeler te maken heb ik van broader gewoon een attribuut gemaakt.

ValenberghsSven commented on app/models/meeting.js:
```diff
@@ -49,8 +47,7 @@ export default Model.extend({
 
   label: computed('plannedStart', 'kind', 'numberRepresentation', function() {
```
computed lijkt niet meer gebruikt? Ik zoek even verder.

ValenberghsSven replied:
op `meeting.label` of `meeting.get('label)` vond ik niks.
`get('label` geeft niets gekoppeld rechtstreeks aan een meeting (behalve kind). 
`.label` geeft veel resultaten, waarvan de meeste niet gerelateerd.
`ModelManager` gebruikt labels, maar ik kon niet direct eentje vinden die meetings als models heeft.
Ik had gisteren ook even een console.error in die computed gestoken en eens door wat views gegaan.
Voorlopig niets gevonden maar ik twijfel of deze echt weg mag, mogelijk heb ik iets gemist.
Deze computed depricated maken voorlopig?

ValenberghsSven replied:
aaahhhh, got it
modelSelector
https://github.com/kanselarij-vlaanderen/frontend-kaleidos/blob/742194e2cf590ce98bffd40a85694ae4a441d4bf/app/components/meeting/new-meeting-modal.hbs#L21

ValenberghsSven replied:
We zouden die computed op model kunnen vervangen door lokaal een modelSelector met block? Zie
https://github.com/kanselarij-vlaanderen/frontend-kaleidos/blob/feature/KAS-2430-gebruik-vergaderactiviteit-codelist-uit-backend-om-keuzelijst-op-te-stellen-in-frontend/app/components/utils/model-manager.hbs#L18

Dan kunnen we die computed met proxies eruit werken?
Kan ook in een latere iteratie en voorlopig deprecate maken. Is op zich buiten scope om te refactoren.
Als ik zoiets tegenkom zet ik er een commentje bij, "refactor nodig, gebruikt op 1 plaats"

sergiofenoll replied:
Ik heb die computed prop er uit gerefactored, maar die ModelSelector is een beetje bulky geworden:
```
<Utils::ModelSelector
  data-test-new-meeting-related-main-meeting
  @modelName="meeting"
  @searchField="number-representation"
  @sortField="-planned-start"
  @selectedItems={{this.selectedMainMeeting}}
  @selectModel={{this.selectMainMeeting}}
  as |model|>
    {{model.kind.printLabel}}
    {{t "of"}}
    {{moment-format model.plannedStart "DD-MM-YYYY"}}
    ({{model.numberRepresentation}})
</Utils::ModelSelector>
```

Als ge liever toch de computed property hebt om niet zoveel logica te zetten in de template, dan kan ik de change undo'en.
Heb er ineens ook een bug uitgehaald (zoeken ging niet bij die meeting modelselector)

ValenberghsSven commented on app/pods/application/route.js:
```diff
@@ -34,8 +35,16 @@ export default class ApplicationRoute extends Route {
+      filter: {
+        ':has:priority': true, // There is an ext:MinisterraadType resource called Annex which we don't want to show in dropdowns. It does not have an ext:priority property, so we use that fact to filter it out
+      },
+      sort: 'priority',
```
mogelijk "position", check backend PR

ValenberghsSven commented:
ik heb op deze branch ook een error tegengekomen  (nu bij elke refresh op agendas overzicht)
nog niet zeker of deze bug ook op dev is.
![image](https://user-images.githubusercontent.com/22245223/160872057-b75dfba1-41eb-428a-83aa-d52057b35160.png)

sergiofenoll committed:
Add printLabel getter to meeting-kind

sergiofenoll committed:
Remove meeting kind types from utils/config.js

sergiofenoll committed:
Fix typo in meeting kind constants

sergiofenoll committed:
Remove meeting.isAnnex

sergiofenoll committed:
Replace meetingKind.priority with meetingKind.position

sergiofenoll committed:
Remove @hideLabel from Utils::KindSelector

sergiofenoll committed:
Make the Utils::KindSelector default choice clearer

sergiofenoll committed:
Revert usage of this.router.transitionTo to prevent weird routing related errors

sergiofenoll commented:
> ik heb op deze branch ook een error tegengekomen (nu bij elke refresh op agendas overzicht) nog niet zeker of deze bug ook op dev is. ![image](https://user-images.githubusercontent.com/22245223/160872057-b75dfba1-41eb-428a-83aa-d52057b35160.png)

Damn, zie het ook gebeuren. 
Na wat google zag ik gelijkaardige errors voorkomen in [een open issue](https://github.com/emberjs/ember.js/issues/18577), maar wat ze daar doen is niet hetzelfde als wat we hier aant doen zijn met de transitions. Ben dan maar toch gaan kijken naar transition stuff en mijn originele change was op bepaalde routes `this.transitionTo` vervangen door `this.router.transitionTo` (want dat eerste is deprecated), maar na die change te undo'en gaan de errors weg. Ben nog wat verward over waarom dat het geval is, maar het zou nu tenminste moeten werken...

sergiofenoll committed:
Remove unneeded parens

sergiofenoll commented:
All Jenkins tests are green :) http://kal-kastaar.s.redpencil.io:8080/job/kaleidos/job/custom_frontend_backend_combination/179/

sergiofenoll committed:
Refactor out meeting.label computed property

erikap commented on app/adapters/meeting-kind.js:
```diff
@@ -0,0 +1,18 @@
+  shouldBackgroundReloadAll(store, snapshot) {
+    return !store.peekAll(snapshot.type.modelName).length;
+  },
+
```
Deze optimalisatie zou ik er uit laten. We hebben in de backend al een cache, dus als exact dezelfde request meermaals gebeurd, hitten we die cache al.
Als we afstappen van `meeting-kind` als apart type resource (zie ook backend PR), kan deze logica ook niet meer toegepast worden op de adapter voor `concept`.

erikap commented on app/components/utils/kind-selector.js:
```diff
@@ -1,29 +1,32 @@
+/**
+ * @argument {boolean} hideLabel Whether to hide the label of the dropdown
+ * @argument {boolean} initializeEmptyKind Whether to initialize the dropdown to the default meeting kind, "Ministerraad", if no meeting kind was provided
+ * @argument {function} setAction Action to perform once a meeting kind has been selected
```
De action-handler voor het selecteren van een type zou ik aanpassen naar `@onChange` (of `@onSelect`). `setAction` heeft geen betekenis.

erikap commented on app/components/utils/kind-selector.js:
```diff
@@ -1,29 +1,32 @@
-  isLoading: null,
-  hideLabel: null,
+/**
+ * @argument {boolean} hideLabel Whether to hide the label of the dropdown
```
Mag weg?

erikap commented on app/components/utils/kind-selector.js:
```diff
@@ -1,29 +1,32 @@
+  @service store;
+  @tracked kind = null;
+  @tracked options = this.store.peekAll('meeting-kind').sortBy('position'); // Meeting kinds get loaded in the agendas route, so we can just peek them here.
+  @tracked defaultKind = this.options.firstObject;
```
Het lijkt mij niet de verantwoordelijkheid van deze selector-component om te bepalen wat de default is, want die is mogelijk niet op alle plaatsen in de app hetzelfde (als deze component in de toekomst op nog andere plaatsen gebruikt zou gaan worden). Het is de consumer van de component (`NewMeetingModal` en `EditMeetingModal`) die moet bepalen welke value geselecteerd is.

Een component interface als `Utils::KindSelector @selectedKind={{this.kind}} @onChange={{...}}` lijkt mij logischer. Het is de verantwoordelijkheid van de consumer om te zorgen dat `this.kind` de juiste waarde bevat (en dus de gewenste default bij initializatie).

Er moet dan ook geen interne `@tracked kind` poperty meer zijn in de `KindSelector`-component. De `@selectedKind` kan doorgegeven worden aan de `PowerSelect` en bij wijziging moet enkel de `@onChange`-handler aangeroepen worden. De `KindSelector`-component wordt dus in feite niets meer dan een wrapper rond `PowerSelect` die zorgt dat de dropdown met de juiste opties geïnitalizeerd wordt.

erikap commented on app/components/meeting/edit-meeting-modal.js:
```diff
@@ -25,12 +22,7 @@ export default class MeetingEditMeetingComponent extends Component {
-        (minsterraad) => minsterraad.uri === this.selectedKindUri
-      )
-    );
+    this.kind = this.args.meeting.kind;
```
Aangezien `meeting.kind` nu een relatie is, moet deze ge-await worden.

erikap commented on app/components/agenda/agenda-header.hbs:
```diff
@@ -15,13 +15,13 @@
               >
                 -
-                {{await @meeting.kindToShow.label}}
+                {{await @meeting.kind.label}}
```
Deze `await` is niet nodig in de template.

erikap commented on app/pods/application/route.js:
```diff
@@ -36,6 +37,14 @@ export default class ApplicationRoute extends Route {
+    await this.store.query('meeting-kind', {
+      'page[size]': PAGE_SIZE.MEETING_KIND,
+      filter: {
+        ':has:position': true, // There is an ext:MinisterraadType resource called Annex which we don't want to show in dropdowns. It does not have a schema:position property, so we use that fact to filter it out
```
`:has:` filter flag is enkel bedoeld voor relaties, niet voor attributes. Zoals al aangegeven in backend PR kan hier beter gefilterd worden op `:has-no:narrower`.

erikap commented on app/config/config.js:
```diff
@@ -9,7 +9,7 @@ export const PAGE_SIZE = {
   ACTIVITIES: 500,
   CASES: 500,
-
+  MEETING_KIND: 5,
```
Je kan de bestaande page size voor `CODE_LISTS` gebruiken.

erikap commented on app/config/constants.js:
```diff
@@ -90,5 +90,12 @@ export default {
     BVR: 'http://themis.vlaanderen.be/id/concept/regelgeving-type/ea7f5f79-f81c-459b-a0f7-d8e90e2d9b88',
-  }
+  },
+  MEETING_KINDS: {
+    MINISTERRAAD: 'http://kanselarij.vo.data.gift/id/concept/ministerraad-type-codes/A5D6B7A8-2F9C-44B6-B3BE-98D80B426254',
+    ELEKTRONISCHE_PROCEDURE: 'http://kanselarij.vo.data.gift/id/concept/ministerraad-type-codes/406F2ECA-524D-47DC-B889-651893135456',
+    BIJZONDERE_MINISTERRAAD: 'http://kanselarij.vo.data.gift/id/concept/ministerraad-type-codes/7D8E35BE-E5D1-494F-B5F9-51B07875B96F',
+    PVV: 'http://kanselarij.vo.data.gift/id/concept/ministerraad-type-codes/1d16cb70-0ae9-489e-bf97-c74897222e3c',
+    ANNEX: 'http://kanselarij.vo.data.gift/id/concept/ministerraad-type-codes/d36138a9-07f0-4df6-bbf0-abd51a24e4ce',
+  },
```
Hier zou ik enkel de URIs opnemen die effectief gebruikt worden in de frontend om de maintenance in de toekomst te beperken.

erikap commented on app/pods/print-overviews/decisions/controller.js:
```diff
@@ -25,7 +25,7 @@ export default Controller.extend({
   titleTranslationParams: computed('model.createdFor', function() {
     const meeting = this.get('model.createdFor');
-    const kindLabel = meeting.get('kindToShow.altLabel');
+    const kindLabel = meeting.get('kind').get('altLabel') || meeting.get('kind').get('label') || '';
```
`kind.printLabel`?

sergiofenoll committed:
Remove meeting-kinds adapter

sergiofenoll committed:
Add kindPrintLabel helper

sergiofenoll committed:
Remove unneeded await

sergiofenoll committed:
Await meeting.kind relation in EditMeetingModal

sergiofenoll committed:
Add extra properties to concept model

sergiofenoll committed:
Add kindPrintLabel modal

sergiofenoll committed:
Simplify KindSelector

sergiofenoll committed:
Update constants

sergiofenoll committed:
Remove meeting-kind, use concept instead

sergiofenoll committed:
Merge branch 'development' into feature/KAS-2430-gebruik-vergaderactiviteit-codelist-uit-backend-om-keuzelijst-op-te-stellen-in-frontend

sergiofenoll committed:
Try updating Jenkinsfile so it automatically checks out backend branch

sergiofenoll committed:
Fix Jenkinsfile syntax

sergiofenoll committed:
Use correct environment variable for branch name in Jenkinsfile

sergiofenoll committed:
One more try for variable name

sergiofenoll committed:
Try other Jenkinsfile env variables

sergiofenoll committed:
Try using built-in Groovy-script branching in Jenkinsfile

sergiofenoll committed:
Use double quotes instead of single quotes!

erikap commented on app/components/meeting/new-meeting-modal.js:
```diff
@@ -169,12 +174,14 @@ export default class MeetingNewMeetingModal extends Component {
     return agendaitem;
   }
 
+  async isAnnexMeetingKind(kind) {
```
This async function still annoys me a bit because of the complex `await` constructions in the template that are caused by this. The only alternative I currently see is defining a `@tracked`-property `isAnnexMeeting` on this component and update it each time `this.kind` changes. Since `this.kind` only gets updated in the `setKind` action it might simplify the code.

erikap commented:
I only realize now that the edit-meeting-modal can break the meeting data :confused: (not caused by this PR). There is no consideration for annex-meetings like in the new-meeting-modal. If the type of the meeting changes to 'Plan VV' the user does not get an option to select a main meeting. Going to create a new story to cover this.

sergiofenoll committed:
Update KindSelector loading mechanism

Otherwise, if it's still loading the tests will complain because they
try to click the loading text instead of waiting

sergiofenoll committed:
Refactor isAnnexMeetingKind async function with @tracked property

sergiofenoll committed:
Revert changes to Jenkinsfile

Will try to get the Jenkins setup to automatically pull the correct
backend branch some other time.

sergiofenoll committed:
Await kind relationship when in print-overviews

sergiofenoll commented:
Behold, working tests :) http://kal-kastaar.s.redpencil.io:8080/job/kaleidos/job/custom_frontend_backend_combination/183/

ValenberghsSven reviewed (approved):
properkes opgelost met concepts

ValenberghsSven commented:
Ik had na approven nog even twijfel of er niets ging stuk gaan bij legacy (gewoon de verschillende views, geen edits) maar dat werkt precies nog goed.
